File: mip_reference/node-express/lib/mip/signature.js
'use strict';

const mipCrypto = require('./crypto');

// Build the data string to sign
function buildSignatureData(timestamp, path, jsonBody) {
  let data = `${timestamp}${path}`;
  if (jsonBody && jsonBody.length > 0) {
    data += jsonBody;
  }
  return data;
}

// Create a MIP request signature
// Per spec: signature signs "timestamp + path + json_payload"
function signRequest(privateKeyPem, timestamp, path, jsonBody) {
  const data = buildSignatureData(timestamp, path, jsonBody);
  return mipCrypto.sign(privateKeyPem, data);
}

// Verify a MIP request signature
function verifyRequest(publicKeyPem, signature, timestamp, path, jsonBody) {
  const data = buildSignatureData(timestamp, path, jsonBody);
  return mipCrypto.verify(publicKeyPem, signature, data);
}

// Check if timestamp is within acceptable window (+/-5 minutes)
function timestampValid(timestamp, windowSeconds = 300) {
  try {
    const requestTime = new Date(timestamp);
    if (isNaN(requestTime.getTime())) return false;
    const now = new Date();
    return Math.abs(now.getTime() - requestTime.getTime()) <= windowSeconds * 1000;
  } catch {
    return false;
  }
}

module.exports = { signRequest, verifyRequest, timestampValid };

File: mip-ruby/lib/member_interchange_protocol/signature.rb
# frozen_string_literal: true

module MemberInterchangeProtocol
  module Signature
    # Create a MIP request signature
    # Per spec: signature signs "timestamp + path + json_payload"
    def self.sign_request(private_key_pem, timestamp, path, json_body = nil)
      data = build_signature_data(timestamp, path, json_body)
      Crypto.sign(private_key_pem, data)
    end

    # Verify a MIP request signature
    def self.verify_request(public_key_pem, signature, timestamp, path, json_body = nil)
      data = build_signature_data(timestamp, path, json_body)
      Crypto.verify(public_key_pem, signature, data)
    end

    # Check if timestamp is within acceptable window (+/-5 minutes)
    def self.timestamp_valid?(timestamp, window_seconds = 300)
      request_time = Time.iso8601(timestamp)
      now = Time.now
      (now - request_time).abs <= window_seconds
    rescue ArgumentError
      false
    end

    class << self
      private

      def build_signature_data(timestamp, path, json_body)
        data = "#{timestamp}#{path}"
        data += json_body.to_s if json_body && !json_body.empty?
        data
      end
    end
  end
end
